### DISCUSSION TITLE
Cannot import .dxf file

### DISCUSSION THREAD

**Discussion by shearstressed** [robbievanleeuwen/section-properties#185] [Q&A] [answered]
Hello everyone! I am trying to import a custom geometry from SolidWorks to section-properties, but the following error comes up: 

"parent = polygons.pop(0)

IndexError: pop from empty list"

Has this happened to anyone in the past? What software do you use to create a custom geometry from CAD and import it?

Thanks in advance!

**Comment by robbievanleeuwen** [comment]
Hi @shearstressed, are you able to share the dxf file? A bit hard to debug without.

Robbie

**Comment by shearstressed** [comment, marked as answer]
Hi @robbievanleeuwen , you do have a point and I'm sorry! Here is the file attached.
I am sending it in the .txt format, since github doesn't accept .dxf files.

Thanks!
 
[test.txt](https://github.com/robbievanleeuwen/section-properties/files/7949377/test.txt)

**Reply by robbievanleeuwen** [reply]
Thanks @shearstressed, *sectionproperties* uses [cad-to-shapely](https://github.com/aegis1980/cad-to-shapely) to load dxf files. It looks like ``cad-to-shapely`` is looking for splines & polylines in the dxf files it loads and there are none of these in your attached file.

I've had success in the past with exporting dxf files from Rhino by using the 2004 LINES or 2004 POLYLINES profile (see [here](http://docs.mcneel.com/rhino/5/help/en-us/fileio/autocad_dwg_dxf_import_export.htm)). I'm not sure what software you are using but perhaps trying to select a similar profile will help?

Wondering if @aegis1980 or @connorferster have anything to add?

**Reply by shearstressed** [reply]
I am using SolidWorks 2021 SP2 and I've also tried using LibreCAD, which is a free software and I still can't import the shapes. I've also tried to apply the profiles you talked about, but I didn't find any similar settings in both softwares I am using.

Anyway, thanks for your help!

**Reply by aegis1980** [reply]
@shearstressed 
[cad-2-shapely](https://github.com/aegis1980/cad-to-shapelyy) is my doing. 
I was fairly selective on what to convert and what to ignore from DXF files (there is a lot of fluff in file) but your line-only angle ought to work really!
So, I'll take a look and see what's missing...or you could and make a pull request. 
Either way it would be helpful if you raised as in issue in [cad-2-shapely](https://github.com/aegis1980/cad-to-shapely). As @robbievanleeuwen said: not an issue with section properties. 

PS Love the **shearstressed**  moniker BTW ;)

**Reply by connorferster** [reply]
Also, love the "shearstressed" name.

I opened up the test file in FreeCAD and thought I could do a join command on all of the line segments but it errored out, which was unexpected.

I think it might be difficult to expand the functionality to include polygons that may or may not exist in a file built from individual line segments. If the assumption is that only one geometry exists in the file, then the lines can just be joined. However if there is more than one geometry (polygon), which lines define which geometry?

I was reviewing the SolidWorks help page (http://help.solidworks.com/2021/english/SolidWorks/sldworks/r_dxf_dwg_export_options.htm?id=586f401f0df441cca98279df7574d452#Pg0) and saw that there a couple of DXF export options that may be helpful. Specifically, end point merging and spline export options.

Perhaps one reason FreeCAD errored out during the join is because maybe the lines were not all completely connected? If that is a possibility, then it may be worth reviewing the shape in SW and trying to export again (perhaps with the end point merging option on, also).

Just some thoughts. I don't have any experience with SW although I would _hope_ that it does not explode all geometry into line segments upon DXF export. If it does not explode all lines, then maybe there is a way to model it such that it does export export polylines and/or splines (implies by the spline options).

**Reply by shearstressed** [reply]
I really appreciate your inputs and I would love to be able to help you guys develop an even more powerful tool, but unfortunately I don't have enough coding skills to be of service. I am just a humble user of your software (which is already amazing as it is!) and I am aware of the hardships of coding such a thing. So, if you ever happen to solve this issue in future updates, I'll be here to cheer you and thank you for your work!
 
All the best,
shearstressed ;)

**Reply by pedro-brands** [reply]
Hi guys! I love your work. Used it so many times (and I used it for mechanical structural design btw) ! I was able to import a .dxf from inventor through your code. So I would assume, since rhino is from Autodesk, that CAD2Shapely, only works well with some type .dxf from Autodesk Softwares.

**Comment by connorferster** [comment]
Thanks @pedro-brands! Glad to hear you are such an avid user :)

I think Rhino is from an independent company owned by Robert McNeel and his McNeel corporation. A polyline is a polyline and it won’t matter which software made it as long as it’s a polyline (thankfully!)

This situation with SolidWorks is puzzling…I wish I had a copy to play with!
